substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +58,115 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    static class SpeculationReasonInvocationHandler implements InvocationHandler {
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return IsolatedSpeculationReasonEncoding.class.getDeclaredMethod(method.getName(), method.getParameterTypes()).invoke(args);
+        }
+    }
+
+    static class SpeculationReasonEncodingSupplier implements Supplier<Object> {
+        private final Object speculationReasonEncodingProxy;
+
+        SpeculationReasonEncodingSupplier(Object speculationReasonEncodingProxy) {
+            this.speculationReasonEncodingProxy = speculationReasonEncodingProxy;
+        }
+
+        @Override
+        public Object get() {
+            try {
+                return speculationReasonEncodingProxy.getClass().getDeclaredConstructor().newInstance();
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                e.printStackTrace();
+            }
+            return null;
+        }
+    }
+
+    private byte[] getByteArray(SpeculationReason reason) {
+        byte[] bytes = null;
+        if (reason instanceof UnencodedSpeculationReason) {
+            IsolatedSpeculationReasonEncoding encoding = encode((UnencodedSpeculationReason) reason);
+            bytes = encoding.getByteArray();
+        } else {
+            try {
+                Class<?> speculationReasonEncodingClass = Class.forName("jdk.vm.ci.meta.SpeculationLog$SpeculationReasonEncoding");
+                final Object speculationReasonEncodingProxy = Proxy.newProxyInstance(speculationReasonEncodingClass.getClassLoader(),
+                                new Class<?>[]{speculationReasonEncodingClass},
+                                new SpeculationReasonInvocationHandler());
+                IsolatedSpeculationReasonEncoding encoding = (IsolatedSpeculationReasonEncoding) Class.forName("jdk.vm.ci.meta.EncodedSpeculationReason").getDeclaredMethod("encode").invoke(new SpeculationReasonEncodingSupplier(speculationReasonEncodingProxy));
+                bytes = encoding.getByteArray();
+            } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException | ClassNotFoundException e) {
+                e.printStackTrace();
+            }
+        }
+        return bytes;
+    }
+
+    /** Adapted from {@code jdk.vm.ci.meta.EncodedSpeculationReason#encode}. */
+    public IsolatedSpeculationReasonEncoding encode(UnencodedSpeculationReason reason) {
+        IsolatedSpeculationReasonEncoding encoding = null;
```

Comment:
Should we cache the encodings here?

Reply:
Looks like we can only do that using a `WeakIdentityHashMap`, but I wouldn't now unless it proves to have a performance impact.

Reply:
Not sure how to benchmark this. I will be happy to if you provide instructions though :)